app/src/main/java/com/amaze/filemanager/ui/fragments/data/MainFragmentViewModel.kt

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (C) 2014-2021 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
+ * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
+ *
+ * This file is part of Amaze File Manager.
+ *
+ * Amaze File Manager is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.amaze.filemanager.ui.fragments.data
+
+import android.content.SharedPreferences
+import android.os.Bundle
+import androidx.lifecycle.ViewModel
+import com.amaze.filemanager.adapters.data.LayoutElementParcelable
+import com.amaze.filemanager.database.CloudHandler
+import com.amaze.filemanager.file_operations.filesystem.OpenMode
+import com.amaze.filemanager.filesystem.HybridFileParcelable
+import com.amaze.filemanager.ui.fragments.preference_fragments.PreferencesConstants
+import com.amaze.filemanager.utils.DataUtils
+import java.util.*
+
+class MainFragmentViewModel : ViewModel() {
+
+    var currentPath: String? = null
+
+    /** This is not an exact copy of the elements in the adapter  */
+    var listElements: ArrayList<LayoutElementParcelable>? = null
+
+    var fileCount = 0
+    var folderCount: Int = 0
+    var columns: Int = 0
+    var smbPath: String? = null
+    var searchHelper = ArrayList<HybridFileParcelable>()
+    var no = 0
+
+    var sortby = 0
+    var dsort = 0
+    var asc = 0
+    var home: String? = null
+    var selection = false
+    var results: Boolean = false
+    lateinit var openMode: OpenMode
+
+    // defines the current visible tab, default either 0 or 1
+    // private int mCurrentTab;
+    /** For caching the back button  */
+    var back: LayoutElementParcelable? = null
+
+    var dragAndDropPreference = 0
+
+    val dataUtils: DataUtils = DataUtils.getInstance()
```

Comment:
Please don't keep an instance of a Singleton in the ViewModel.

Reply:
Instance isn't kept here.. this is just a reference to the original instance, right. Instance is maintained by dataUtils itself.

Reply:
Sorry, I meant, don't keep a reference to a Singleton, it is already accessible via DataUtils.getInstance().

Reply:
It's used by mainFragment. Either I keep it here or in mainfragment (which isn't good IMO since it's view layer). We'll slowly keep migrating to viewmodels and remove DataUtils singleton alltogether.

Reply:
Can't you replace all calls for `MainFragmentViewModel.getDataUtils()` with `DataUtils.getInstance()`?

Reply:
I just don't see the issue here. We can have soft references of singleton without any problems afaik. Regardless, I've made the change.